service/src/main/kotlin/fi/espoo/evaka/reservations/ReservationQueries.kt

```diff
@@ -542,3 +546,224 @@ fun Database.Read.getReservationContractDayRanges(
+                      FROM unnest(:confirmedDays) AS t(date)) day
+                         LEFT JOIN backup_care bc ON daterange(bc.start_date, bc.end_date, '[]') @> day.date
+                    AND bc.unit_id = :unitId)
+      SELECT pcd.date,
+             pcd.group_id,
+             CASE
+                 WHEN p.id IS NULL
+                     THEN null
+                 ELSE CASE
+                          WHEN (ab.absence_type IS NOT NULL OR bc.unit_id IS NOT NULL)
+                              THEN 'ABSENT'
+                          ELSE 'PRESENT'
+                          END
+                 END as absence_status,
+             CASE
+                 WHEN u.type && array ['FAMILY', 'GROUP_FAMILY']::care_types[]
+                     THEN $familyUnitPlacementCoefficient
+                 ELSE CASE
+                          WHEN (extract('year' from age(pcd.date, p.date_of_birth)) < 3)
+                              THEN coalesce(
+                                  sno.realized_occupancy_coefficient_under_3y,
+                                  sno_default.realized_occupancy_coefficient_under_3y)
+                          ELSE coalesce(
+                                  sno.realized_occupancy_coefficient,
+                                  sno_default.realized_occupancy_coefficient)
+                     END
+                 END *
+             coalesce(af.capacity_factor, 1.00)  as occupancy_coefficient
+      FROM placement_days pcd
+               JOIN daycare u ON u.id = :unitId
+               LEFT JOIN absence ab
+                         ON ab.child_id = pcd.child_id
+                             AND ab.date = pcd.date
+               LEFT JOIN backup_care bc
+                         ON bc.child_id = pcd.child_id
+                             AND bc.unit_id <> :unitId
+                             AND
+                            daterange(bc.start_date, bc.end_date, '[]') @>
+                            pcd.date
+               LEFT JOIN placement pl
+                         ON pcd.child_id = pl.child_id
+                             AND
+                            daterange(pl.start_date, pl.end_date, '[]') @>
+                            pcd.date
+               LEFT JOIN service_need sn
+                         ON pl.id = sn.placement_id
+               LEFT JOIN service_need_option sno
+                         ON sn.option_id = sno.id
+               LEFT JOIN service_need_option sno_default
+                         ON sno_default.default_option IS TRUE
+                             AND
+                            sno_default.valid_placement_type =
+                            pl.type
+               LEFT JOIN person p
+                         ON pcd.child_id = p.id
+               LEFT JOIN assistance_factor af
+                         ON pcd.child_id = af.child_id
+                             AND
+                            af.valid_during @> pcd.date) a
+group by a.date, a.group_id
```

Comment:
Overall, the totally different code style in this query compared to other queries in the code base hurts the readability a lot.

Reply:
Query refactored around suggested db function.